### ISSUE TITLE
iothub_client: valgrind reports uninitialised values

### ISSUE THREAD

**Issue by jdswensen** [Azure/azure-iot-sdk-c#360]
- **Hardware**
Raspberry Pi 3, but I have noted this behavior on other ARM based boards.

- **OS and version used:** <VERSION> 
```
pi@dev0:~ $ uname -a
Linux dev0 4.9.59-v7+ #1047 SMP Sun Oct 29 12:19:23 GMT 2017 armv7l GNU/Linux
pi@dev0:~ $ cat /etc/os-release 
PRETTY_NAME="Raspbian GNU/Linux 9 (stretch)"
NAME="Raspbian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

- **SDK version used:** <VERSION> 
Commit hash: `4ec2d11d004df586b9f25b54e31c63cdd449dc31`

- **GCC version:**
```
pi@dev0:~ $ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/arm-linux-gnueabihf/6/lto-wrapper
Target: arm-linux-gnueabihf
Configured with: ../src/configure -v --with-pkgversion='Raspbian 6.3.0-18+rpi1' --with-bugurl=file:///usr/share/doc/gcc-6/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-6 --program-prefix=arm-linux-gnueabihf- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-libitm --disable-libquadmath --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-6-armhf/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-6-armhf --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-6-armhf --with-arch-directory=arm --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-sjlj-exceptions --with-arch=armv6 --with-fpu=vfp --with-float=hard --enable-checking=release --build=arm-linux-gnueabihf --host=arm-linux-gnueabihf --target=arm-linux-gnueabihf
Thread model: posix
gcc version 6.3.0 20170516 (Raspbian 6.3.0-18+rpi1) 

pi@dev0:~ $ apt-cache policy gcc
gcc:
  Installed: 4:6.3.0-4
  Candidate: 4:6.3.0-4
  Version table:
 *** 4:6.3.0-4 500
        500 http://mirrordirector.raspbian.org/raspbian stretch/main armhf Packages
        100 /var/lib/dpkg/status

```

- **Openssl version:**
```
pi@dev0:~ $ openssl version -a
OpenSSL 1.1.0f  25 May 2017
built on: reproducible build, date unspecified
platform: debian-armhf
compiler: gcc -DDSO_DLFCN -DHAVE_DLFCN_H -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/lib/ssl\"" -DENGINESDIR="\"/usr/lib/arm-linux-gnueabihf/engines-1.1\"" 
OPENSSLDIR: "/usr/lib/ssl"
ENGINESDIR: "/usr/lib/arm-linux-gnueabihf/engines-1.1"

pi@dev0:~ $ apt-cache policy openssl
openssl:
  Installed: 1.1.0f-3+deb9u1
  Candidate: 1.1.0f-3+deb9u1
  Version table:
 *** 1.1.0f-3+deb9u1 500
        500 http://mirrordirector.raspbian.org/raspbian stretch/main armhf Packages
        100 /var/lib/dpkg/status
```

- **Installed library versions:**
```
315 libs found in cache `/etc/ld.so.cache'
	libz.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libz.so.1
	libz.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libz.so
	libxtables.so.12 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libxtables.so.12
	libxml2.so.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libxml2.so.2
	libxcb.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libxcb.so.1
	libxapian.so.30 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libxapian.so.30
	libwrap.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libwrap.so.0
	libwbclient.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libwbclient.so.0
	libv4l2rds.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libv4l2rds.so.0
	libv4l2.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libv4l2.so.0
	libv4l1.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libv4l1.so.0
	libv4lconvert.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libv4lconvert.so.0
	libvcsm.so (libc6,hard-float) => /opt/vc/lib/libvcsm.so
	libvcos.so (libc6,hard-float) => /opt/vc/lib/libvcos.so
	libvchiq_arm.so (libc6,hard-float) => /opt/vc/lib/libvchiq_arm.so
	libuv.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libuv.so.1
	libuuid.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libuuid.so.1
	libuuid.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libuuid.so
	libutil.so.1 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libutil.so.1
	libutil.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libutil.so
	libustr-1.0.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libustr-1.0.so.1
	libusb-1.0.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libusb-1.0.so.0
	libusb-0.1.so.4 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libusb-0.1.so.4
	libunistring.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libunistring.so.0
	libudev.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libudev.so.1
	libudev.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libudev.so.0
	libubsan.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libubsan.so.0
	libtirpc.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libtirpc.so.1
	libtinfo.so.5 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libtinfo.so.5
	libtic.so.5 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libtic.so.5
	libthread_db.so.1 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libthread_db.so.1
	libthread_db.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libthread_db.so
	libtasn1.so.6 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libtasn1.so.6
	libtalloc.so.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libtalloc.so.2
	libsystemd.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libsystemd.so.0
	libsysfs.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libsysfs.so.2
	libstdc++.so.6 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libstdc++.so.6
	libssl.so.1.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libssl.so.1.1
	libssl.so.1.0.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libssl.so.1.0.2
	libssl.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libssl.so
	libssh2.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libssh2.so.1
	libss.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libss.so.2
	libsqlite3.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libsqlite3.so.0
	libsqlite3.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libsqlite3.so
	libsqlite.so.0 (libc6,hard-float) => /usr/lib/libsqlite.so.0
	libsqlite.so (libc6,hard-float) => /usr/lib/libsqlite.so
	libsmartcols.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libsmartcols.so.1
	libslang.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libslang.so.2
	libsigc-2.0.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libsigc-2.0.so.0
	libsigc-1.2.so.5 (libc6,hard-float) => /usr/lib/libsigc-1.2.so.5
	libserializer.so.1 (libc6,hard-float) => /usr/local/lib/libserializer.so.1
	libserializer.so (libc6,hard-float) => /usr/local/lib/libserializer.so
	libsepol.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libsepol.so.1
	libsemanage.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libsemanage.so.1
	libselinux.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libselinux.so.1
	libseccomp.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libseccomp.so.2
	libsasl2.so.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libsasl2.so.2
	libsamplerate.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libsamplerate.so.0
	librtmp.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/librtmp.so.1
	librt.so.1 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/librt.so.1
	librt.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/librt.so
	libresolv.so.2 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libresolv.so.2
	libresolv.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libresolv.so
	libreadline.so.7 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libreadline.so.7
	libreadline.so.6 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libreadline.so.6
	libp11-kit.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libp11-kit.so.0
	libpython3.5m.so.1.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libpython3.5m.so.1.0
	libpthread.so.0 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libpthread.so.0
	libpsl.so.5 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libpsl.so.5
	libprocps.so.6 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libprocps.so.6
	libpopt.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libpopt.so.0
	libpng16.so.16 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libpng16.so.16
	libpng16.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libpng16.so
	libply.so.4 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libply.so.4
	libply-splash-graphics.so.4 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libply-splash-graphics.so.4
	libply-splash-core.so.4 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libply-splash-core.so.4
	libply-boot-client.so.4 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libply-boot-client.so.4
	libpipeline.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libpipeline.so.1
	libperl.so.5.24 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libperl.so.5.24
	libpcsclite.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libpcsclite.so.1
	libpcreposix.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libpcreposix.so.3
	libpcre.so.3 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libpcre.so.3
	libpcprofile.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libpcprofile.so
	libparted.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libparted.so.2
	libpanelw.so.5 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libpanelw.so.5
	libpanel.so.5 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libpanel.so.5
	libpamc.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libpamc.so.0
	libpam_misc.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libpam_misc.so.0
	libpam.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libpam.so.0
	libopenmaxil.so (libc6,hard-float) => /opt/vc/lib/libopenmaxil.so
	libopcodes-2.28-system.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libopcodes-2.28-system.so
	libnss_nisplus.so.2 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libnss_nisplus.so.2
	libnss_nisplus.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libnss_nisplus.so
	libnss_nis.so.2 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libnss_nis.so.2
	libnss_nis.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libnss_nis.so
	libnss_mdns6_minimal.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libnss_mdns6_minimal.so.2
	libnss_mdns6.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libnss_mdns6.so.2
	libnss_mdns4_minimal.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libnss_mdns4_minimal.so.2
	libnss_mdns4.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libnss_mdns4.so.2
	libnss_mdns_minimal.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libnss_mdns_minimal.so.2
	libnss_mdns.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libnss_mdns.so.2
	libnss_hesiod.so.2 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libnss_hesiod.so.2
	libnss_hesiod.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libnss_hesiod.so
	libnss_files.so.2 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libnss_files.so.2
	libnss_files.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libnss_files.so
	libnss_dns.so.2 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libnss_dns.so.2
	libnss_dns.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libnss_dns.so
	libnss_compat.so.2 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libnss_compat.so.2
	libnss_compat.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libnss_compat.so
	libnsl.so.1 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libnsl.so.1
	libnsl.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libnsl.so
	libnpth.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libnpth.so.0
	libnl-3.so.200 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libnl-3.so.200
	libnl-genl-3.so.200 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libnl-genl-3.so.200
	libnih.so.1 (libc6,hard-float) => /lib/libnih.so.1
	libnih-dbus.so.1 (libc6,hard-float) => /lib/libnih-dbus.so.1
	libnghttp2.so.14 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libnghttp2.so.14
	libnfsidmap.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libnfsidmap.so.0
	libnfnetlink.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libnfnetlink.so.0
	libnewt.so.0.52 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libnewt.so.0.52
	libnettle.so.6 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libnettle.so.6
	libnetfilter_conntrack.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libnetfilter_conntrack.so.3
	libncursesw.so.5 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libncursesw.so.5
	libncurses.so.5 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libncurses.so.5
	libmtp.so.9 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libmtp.so.9
	libmpfr.so.4 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libmpfr.so.4
	libmpdec.so.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libmpdec.so.2
	libmpc.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libmpc.so.3
	libmount.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libmount.so.1
	libmnl.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libmnl.so.0
	libmnl.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libmnl.so
	libmmal_vc_client.so (libc6,hard-float) => /opt/vc/lib/libmmal_vc_client.so
	libmmal_util.so (libc6,hard-float) => /opt/vc/lib/libmmal_util.so
	libmmal_core.so (libc6,hard-float) => /opt/vc/lib/libmmal_core.so
	libmmal_components.so (libc6,hard-float) => /opt/vc/lib/libmmal_components.so
	libmmal.so (libc6,hard-float) => /opt/vc/lib/libmmal.so
	libmenuw.so.5 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libmenuw.so.5
	libmenu.so.5 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libmenu.so.5
	libmemusage.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libmemusage.so
	libmagic.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libmagic.so.1
	libm.so.6 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libm.so.6
	libm.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libm.so
	liblz4.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/liblz4.so.1
	liblzo2.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/liblzo2.so.2
	liblzma.so.5 (libc6,hard-float) => /lib/arm-linux-gnueabihf/liblzma.so.5
	liblwres.so.141 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/liblwres.so.141
	liblognorm.so.5 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/liblognorm.so.5
	liblogging-stdlog.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/liblogging-stdlog.so.0
	libldap_r-2.4.so.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libldap_r-2.4.so.2
	liblber-2.4.so.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/liblber-2.4.so.2
	libk5crypto.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libk5crypto.so.3
	libksba.so.8 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libksba.so.8
	libkrb5support.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libkrb5support.so.0
	libkrb5.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libkrb5.so.3
	libkmod.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libkmod.so.2
	libkeyutils.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libkeyutils.so.1
	libjsoncpp.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libjsoncpp.so.1
	libjpeg.so.62 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libjpeg.so.62
	libjim.so.0.76 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libjim.so.0.76
	libjansson.so.4 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libjansson.so.4
	libjansson.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libjansson.so
	libiw.so.30 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libiw.so.30
	libisl.so.15 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libisl.so.15
	libisccfg.so.140 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libisccfg.so.140
	libisccc.so.140 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libisccc.so.140
	libisc.so.160 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libisc.so.160
	libisc-pkcs11.so.160 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libisc-pkcs11.so.160
	libisc-export.so.160 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libisc-export.so.160
	libip6tc.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libip6tc.so.0
	libip4tc.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libip4tc.so.0
	libiptc.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libiptc.so.0
	libiothub_service_client.so.1 (libc6,hard-float) => /usr/local/lib/libiothub_service_client.so.1
	libiothub_service_client.so (libc6,hard-float) => /usr/local/lib/libiothub_service_client.so
	libiothub_client_mqtt_ws_transport.so.1 (libc6,hard-float) => /usr/local/lib/libiothub_client_mqtt_ws_transport.so.1
	libiothub_client_mqtt_ws_transport.so (libc6,hard-float) => /usr/local/lib/libiothub_client_mqtt_ws_transport.so
	libiothub_client_mqtt_transport.so.1 (libc6,hard-float) => /usr/local/lib/libiothub_client_mqtt_transport.so.1
	libiothub_client_mqtt_transport.so (libc6,hard-float) => /usr/local/lib/libiothub_client_mqtt_transport.so
	libiothub_client_http_transport.so.1 (libc6,hard-float) => /usr/local/lib/libiothub_client_http_transport.so.1
	libiothub_client_http_transport.so (libc6,hard-float) => /usr/local/lib/libiothub_client_http_transport.so
	libiothub_client_amqp_ws_transport.so.1 (libc6,hard-float) => /usr/local/lib/libiothub_client_amqp_ws_transport.so.1
	libiothub_client_amqp_ws_transport.so (libc6,hard-float) => /usr/local/lib/libiothub_client_amqp_ws_transport.so
	libiothub_client_amqp_transport.so.1 (libc6,hard-float) => /usr/local/lib/libiothub_client_amqp_transport.so.1
	libiothub_client_amqp_transport.so (libc6,hard-float) => /usr/local/lib/libiothub_client_amqp_transport.so
	libidn2.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libidn2.so.0
	libidn.so.11 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libidn.so.11
	libident.so.0 (libc6,hard-float) => /usr/lib/libident.so.0
	libicuuc.so.57 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libicuuc.so.57
	libicutu.so.57 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libicutu.so.57
	libicutest.so.57 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libicutest.so.57
	libiculx.so.57 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libiculx.so.57
	libicule.so.57 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libicule.so.57
	libicui18n.so.57 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libicui18n.so.57
	libicuio.so.57 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libicuio.so.57
	libicudata.so.57 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libicudata.so.57
	libhogweed.so.4 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libhogweed.so.4
	libhistory.so.7 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libhistory.so.7
	libhistory.so.6 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libhistory.so.6
	libgthread-2.0.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libgthread-2.0.so.0
	libgssapi_krb5.so.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libgssapi_krb5.so.2
	libgpg-error.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libgpg-error.so.0
	libgomp.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libgomp.so.1
	libgobject-2.0.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libgobject-2.0.so.0
	libgnutls.so.30 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libgnutls.so.30
	libgmp.so.10 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libgmp.so.10
	libgmodule-2.0.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libgmodule-2.0.so.0
	libglib-2.0.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libglib-2.0.so.0
	libgio-2.0.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libgio-2.0.so.0
	libgdbm_compat.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libgdbm_compat.so.3
	libgdbm.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libgdbm.so.3
	libgcrypt.so.20 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libgcrypt.so.20
	libgcc_s.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libgcc_s.so.1
	libfreetype.so.6 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libfreetype.so.6
	libfreetype.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libfreetype.so
	libformw.so.5 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libformw.so.5
	libform.so.5 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libform.so.5
	libfftw3f_threads.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libfftw3f_threads.so.3
	libfftw3f_omp.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libfftw3f_omp.so.3
	libfftw3f.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libfftw3f.so.3
	libffi.so.6 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libffi.so.6
	libfdisk.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libfdisk.so.1
	libfastjson.so.4 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libfastjson.so.4
	libfakeroot-0.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libfakeroot/libfakeroot-0.so
	libe2p.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libe2p.so.2
	libext2fs.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libext2fs.so.2
	libexpatw.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libexpatw.so.1
	libexpat.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libexpat.so.1
	libevent-2.0.so.5 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libevent-2.0.so.5
	libestr.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libestr.so.0
	libelftoolchain.so (libc6,hard-float) => /opt/vc/lib/libelftoolchain.so
	libelf.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libelf.so.1
	libedit.so.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libedit.so.2
	libdw.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libdw.so.1
	libdtovl.so (libc6,hard-float) => /opt/vc/lib/libdtovl.so
	libdrm.so.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libdrm.so.2
	libdns.so.162 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libdns.so.162
	libdns-pkcs11.so.162 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libdns-pkcs11.so.162
	libdns-export.so.162 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libdns-export.so.162
	libdl.so.2 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libdl.so.2
	libdl.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libdl.so
	libdevmapper.so.1.02.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libdevmapper.so.1.02.1
	libdebug_sym.so (libc6,hard-float) => /opt/vc/lib/libdebug_sym.so
	libdebconfclient.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libdebconfclient.so.0
	libdbus-1.so.3 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libdbus-1.so.3
	libdb-5.3.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libdb-5.3.so
	libdaemon.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libdaemon.so.0
	libcwidget.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libcwidget.so.3
	libcurl.so.4 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libcurl.so.4
	libcurl.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libcurl.so
	libcurl-gnutls.so.4 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libcurl-gnutls.so.4
	libcryptsetup.so.4 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libcryptsetup.so.4
	libcrypto.so.1.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libcrypto.so.1.1
	libcrypto.so.1.0.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libcrypto.so.1.0.2
	libcrypto.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libcrypto.so
	libcrypt.so.1 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libcrypt.so.1
	libcrypt.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libcrypt.so
	libcontainers.so (libc6,hard-float) => /opt/vc/lib/libcontainers.so
	libcom_err.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libcom_err.so.2
	libcidn.so.1 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libcidn.so.1
	libcidn.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libcidn.so
	libcc1.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libcc1.so.0
	libcap.so.2 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libcap.so.2
	libcap-ng.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libcap-ng.so.0
	libc.so.6 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libc.so.6
	libbz2.so.1.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libbz2.so.1.0
	libbsd.so.0 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libbsd.so.0
	libbsd.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libbsd.so
	libbrcmWFC.so (libc6,hard-float) => /opt/vc/lib/libbrcmWFC.so
	libbrcmOpenVG.so (libc6,hard-float) => /opt/vc/lib/libbrcmOpenVG.so
	libbrcmGLESv2.so (libc6,hard-float) => /opt/vc/lib/libbrcmGLESv2.so
	libbrcmEGL.so (libc6,hard-float) => /opt/vc/lib/libbrcmEGL.so
	libboost_system.so.1.62.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libboost_system.so.1.62.0
	libboost_iostreams.so.1.62.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libboost_iostreams.so.1.62.0
	libboost_iostreams.so.1.60.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libboost_iostreams.so.1.60.0
	libboost_iostreams.so.1.58.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libboost_iostreams.so.1.58.0
	libboost_filesystem.so.1.62.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libboost_filesystem.so.1.62.0
	libblkid.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libblkid.so.1
	libbind9.so.140 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libbind9.so.140
	libbfd-2.28-system.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libbfd-2.28-system.so
	libbcm_host.so (libc6,hard-float) => /opt/vc/lib/libbcm_host.so
	libbabeltrace.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libbabeltrace.so.1
	libbabeltrace-lttng-live.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libbabeltrace-lttng-live.so.1
	libbabeltrace-dummy.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libbabeltrace-dummy.so.1
	libbabeltrace-ctf.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libbabeltrace-ctf.so.1
	libbabeltrace-ctf-text.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libbabeltrace-ctf-text.so.1
	libbabeltrace-ctf-metadata.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libbabeltrace-ctf-metadata.so.1
	libavahi-core.so.7 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libavahi-core.so.7
	libavahi-common.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libavahi-common.so.3
	libaudit.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libaudit.so.1
	libattr.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libattr.so.1
	libatomic.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libatomic.so.1
	libassuan.so.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libassuan.so.0
	libasound.so.2 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libasound.so.2
	libasan.so.3 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libasan.so.3
	libarmmem.so (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libarmmem.so
	libarchive.so.13 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libarchive.so.13
	libapt-private.so.0.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libapt-private.so.0.0
	libapt-pkg.so.5.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libapt-pkg.so.5.0
	libapt-inst.so.2.0 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libapt-inst.so.2.0
	libapparmor.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libapparmor.so.1
	libanl.so.1 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libanl.so.1
	libanl.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libanl.so
	libacl.so.1 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libacl.so.1
	libX11.so.6 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libX11.so.6
	libXmuu.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libXmuu.so.1
	libXext.so.6 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libXext.so.6
	libXdmcp.so.6 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libXdmcp.so.6
	libXau.so.6 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libXau.so.6
	libSegFault.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libSegFault.so
	libGeoIP.so.1 (libc6,hard-float) => /usr/lib/arm-linux-gnueabihf/libGeoIP.so.1
	libBrokenLocale.so.1 (libc6,hard-float, OS ABI: Linux 3.2.0) => /lib/arm-linux-gnueabihf/libBrokenLocale.so.1
	libBrokenLocale.so (libc6,hard-float, OS ABI: Linux 3.2.0) => /usr/lib/arm-linux-gnueabihf/libBrokenLocale.so
	ld-linux-armhf.so.3 (libc6,hard-float) => /lib/arm-linux-gnueabihf/ld-linux-armhf.so.3
	ld-linux-armhf.so.3 (libc6,hard-float) => /lib/ld-linux-armhf.so.3
```

# Description of the issue:

Performed a native compile on the RPI3. Sample application for library outputs valgrind memory issues which may indicate leaks within the core SDK (not just the sample). Sample application also hangs and does not respond to `SIGINT` while running with valgrind.

# Code sample exhibiting the issue:

[iothub_client_sample_amqp_websockets](https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/samples/iothub_client_sample_amqp_websockets/iothub_client_sample_amqp_websockets.c)

# Console log of the issue:

`valgrind --log-file="sample-valgrind.txt" ./iothub_client_sample_amqp_websockets`

sample-valgrind.txt:
```
==5400== Memcheck, a memory error detector
==5400== Copyright (C) 2002-2011, and GNU GPL'd, by Julian Seward et al.
==5400== Using Valgrind-3.7.0 and LibVEX; rerun with -h for copyright info
==5400== Command: ./iothub_client_sample_amqp_websockets
==5400== Parent PID: 706
==5400== 
--5400-- WARNING: Serious error when reading debug info
--5400-- When reading debug info from /lib/arm-linux-gnueabihf/ld-2.24.so:
--5400-- Ignoring non-Dwarf2/3/4 block in .debug_info
--5400-- WARNING: Serious error when reading debug info
--5400-- When reading debug info from /lib/arm-linux-gnueabihf/ld-2.24.so:
--5400-- Last block truncated in .debug_info; ignoring
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401B230: index (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401B234: index (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x4008B8C: expand_dynamic_string_token (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401B6E0: strlen (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401B6E4: strlen (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x40194AC: __libc_memalign (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x40194B8: __libc_memalign (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C0F8: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C118: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C128: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C144: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C14C: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C174: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C174: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C19C: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C1B0: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C1B4: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C1C0: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x40195EC: calloc (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x400C6D4: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C6EC: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C140: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C704: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C724: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C7A0: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C7B4: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C880: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C7F0: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x400C930: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C104: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C8E4: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C900: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C870: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x40064AC: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x40064D8: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x40064DC: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x4006578: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x40065BC: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x40065C0: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x4006748: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x40067D4: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x40067EC: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x40067F8: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x4007098: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x40068B8: _dl_map_object_from_fd (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x40124A0: _dl_name_match_p (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x40124B4: _dl_name_match_p (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401B280: strcmp (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x40124A8: _dl_name_match_p (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401B678: strdup (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C170: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C170: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C198: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x40098F8: _dl_map_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Syscall param open(filename) contains uninitialised byte(s)
==5400==    at 0x401ABBC: open (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x400C70C: _dl_new_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401B690: strlen (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401B6A8: strlen (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C138: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C178: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C178: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C1A0: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x4009CBC: _dl_map_object (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C2B0: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C1BC: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C1B8: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C1BC: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Conditional jump or move depends on uninitialised value(s)
==5400==    at 0x401C1C4: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C410: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C1A4: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C1C0: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
==5400== Use of uninitialised value of size 4
==5400==    at 0x401C1C4: memcpy (in /lib/arm-linux-gnueabihf/ld-2.24.so)
==5400== 
```

# Consequences of issue:
Pre-existing valgrind issues make it difficult to use the SDK in custom embedded systems applications when trying to reduce memory errors.

**Comment by jspaith** [comment]
@jdswensen - thanks for feedback here.

My first thought is that the samples aren't really designed to be "ready-to-ship".  We're investigating making this clearer with some code changes; will still have samples obviously.  On this sample in particular, one possibility is the context passed in `IoTHubClient_SendEventAsync` can be accessed on both the main() thread and also the `SendConfirmationCallback` without a lock being held.  Even on stuff like simple atomic read / write which are safe, I've seen Valgrind get unhappy if there's not locks.

We have e2e tests for AMQP_WS in `$iothub_client\tests\iothubclient_amqp_ws_e2e` and all our checkins are gated by Valgrind - we can't merge to main if anything fails.  We don't run this on a Pi but it is on an Ubuntu which should be pretty close in terms of platform abstraction layer.

When we run Valrgind, we also run with a suppressions file with some of false positives we've identified over time.  $\iothub_client\tests\global_valgrind_suppression.supp.  I'd recommend feeding that into any Valgrind runs so you don't have noise.

It's great you're doing Valgrind - wish more folks did.  If you still hit issues, let us know.  There's not a lot to go on in the stacks above, so i don't know if bumping --num-callers would help or if getting symbols lined up would, but if this still repros on product code we'd need a deeper stack to understand what's up.

**Comment by jdswensen** [comment]
@jspaith - Here is some more detail for you.

I'm working on integrating this SDK into our embedded product which runs on an iMX6UL (ARMv7). The build tool we use is called [Yocto](https://www.yoctoproject.org/). I'm also working with Mark Radbourne on [issue 229](https://github.com/Azure/azure-iot-sdk-c/issues/229). When I put the build system, Azure SDK patch, and target application in place, valgrind blew up. After 1000 errors, it informed me it was giving up. I started decoupling different parts of the process and narrowed it down to the Azure SDK. I know the samples aren't meant to be commercialized out of the box, but I think there may be issues in the core SDK that I'm unable to diagnose. I ran through this again with the suppression file and `--num-callers=24` with the same results.

I thought the Pi might be enough to go on, but after some digging, I'm not sure the Raspian version of valgrind is properly reporting errors because of this bug:

```
--5400-- WARNING: Serious error when reading debug info
--5400-- When reading debug info from /lib/arm-linux-gnueabihf/ld-2.24.so:
--5400-- Ignoring non-Dwarf2/3/4 block in .debug_info
--5400-- WARNING: Serious error when reading debug info
--5400-- When reading debug info from /lib/arm-linux-gnueabihf/ld-2.24.so:
--5400-- Last block truncated in .debug_info; ignoring
```

The compressed debug info [bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=810295) was fixed in valgrind v1.3.12, but Raspian uses an older version. I still think there are valid errors here, valgrind just doesn't know how to interpret them because of the compressed debug info.

So I switched back to using Yocto to build everything. This time, I pointed the build system to commit `4f5b1f4fcafba68a93f322a2c34bd4d064a93750`. I ran the command `valgrind --suppressions=./global_valgrind_suppression.supp --num-callers=24 --log-file="sample-valgrind-imx6ul.log
[sample-valgrind-imx6ul.log](https://github.com/Azure/azure-iot-sdk-c/files/1675576/sample-valgrind-imx6ul.log)
" ./iothub_client_sample_amqp_websockets` on our target. I attached the log file because it is way too long to post inline.

Right now, I'm looking into ways to replicate this on your end since you don't have access to my hardware. The thing I'm most concerned about is architectural differences between ARM and x86. Does your SDK build process assume all linux builds will be x86? If that's the case, are you relying on any x86 specific instructions, designs, or behaviors that don't exist on ARM? 

My secondary concern would be library versions. Are you relying on any new-ish features in curl, libcrypto, libssl, etc? Do you state the minimum required versions of these libraries anywhere?

The good news is that valgrind isn't reporting any definitely, indirectly, or possibly lost bytes. Our application still runs, unfortunately the errors reported due to the Azure SDK make checking for leaks impossible.

--Jake
[sample-valgrind-imx6ul.log](https://github.com/Azure/azure-iot-sdk-c/files/1675592/sample-valgrind-imx6ul.log)

**Comment by jdswensen** [comment]
I also took a look at the E2E tests. I'm getting a run-time error about config files. Is there a doc or walkthough on configuring and running them, or are they only part of your internal process?

**Comment by jspaith** [comment]
I've looked at E2E test setup and it's theoretically possible for you to run on your own since they just consume environment variables to your Hub, but... it's a complex setup.  As you'll see below i don't think it'll help here so it's not worth it for you.

I don't know of recommended library versions, will update if I learn anything.

For the meat of the issue here RE Valgrind - 1st thanks for the much better stacks.  There are not ARM/x86 differences in the code.  Even if we did, they'd be hidden in a PAL and not necessarily at this layer.  And FWIW the dependencies you're using are what we use in our Ubuntu Valgrind run.

My gut - to be validated - is that this is either a false positive or... there's a difference in timing on your board versus our tests.  The code to build up this buffer is complex, so even though it's single-threaded if your device was giving us say 5 bytes at a time and our tests were giving us 500, it's possible there's something off there in the 5 byte case.  This is still speculation and I didn't see any obvious problems.  I'll need to be doing some CR or trying to instrument.  Will let you know if I haev more theories or need more data.

**Comment by jspaith** [comment]
Some interesting lines in your log

```
disInstr(arm): unhandled instruction: 0xEC510F1E
                 cond=14(0xE) 27:20=197(0xC5) 4:4=1 3:0=14(0xE)
==519== valgrind: Unrecognised instruction at address 0x4a38848.
==519==    at 0x4A38848: ??? (in /usr/lib/libcrypto.so.1.0.2)
==519== Your program just tried to execute an instruction that Valgrind
==519== did not recognise.  There are two possible reasons for this.
==519== 1. Your program has a bug and erroneously jumped to a non-code
==519==    location.  If you are running Memcheck and you just saw a
==519==    warning about a bad jump, it's probably your program's fault.
==519== 2. The instruction is legitimate but Valgrind doesn't handle it,
==519==    i.e. it's Valgrind's fault.  If you think this is the case or
==519==    you are not sure, please let us know and we'll try to fix it.
==519== Either way, Valgrind will now raise a SIGILL signal which will
==519== probably kill your program.
```

I did some research and found https://bugs.kde.org/show_bug.cgi?id=344802.  Could you try the "--sigill-diagnostics=no" option the post recommends.  Obviously the exe keeps running but this is definitely a flag, especially as warnings are closely tied to the SSL/crypto layer.

**Comment by jdswensen** [comment]
It looks like that suppressed some of the output. I'm still getting an "uninitialised value(s)" error though. 

```
==1191== Memcheck, a memory error detector
==1191== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==1191== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==1191== Command: ./iothub_client_sample_amqp_websockets
==1191== Parent PID: 499
==1191== 
==1191== Thread 2:
==1191== Conditional jump or move depends on uninitialised value(s)
==1191==    at 0x4014398: dl_open_worker (dl-open.c:202)
==1191==    by 0x4C6B26B: _dl_catch_error (dl-error-skeleton.c:198)
==1191==    by 0x4013E93: _dl_open (dl-open.c:649)
==1191==    by 0x4C6A86B: do_dlopen (dl-libc.c:87)
==1191==    by 0x4C6B26B: _dl_catch_error (dl-error-skeleton.c:198)
==1191==    by 0x4C6A937: dlerror_run (dl-libc.c:46)
==1191==    by 0x4C6A9AF: __libc_dlopen_mode (dl-libc.c:163)
==1191==    by 0x4C5402F: nss_load_library (nsswitch.c:358)
==1191==    by 0x4C5489F: __nss_lookup_function (nsswitch.c:455)
==1191==    by 0x4C549D3: __nss_lookup (nsswitch.c:189)
==1191==    by 0x4C445BF: gethostbyname2_r@@GLIBC_2.4 (getXXbyYY_r.c:262)
==1191==    by 0x4C1B54B: gaih_inet.constprop.6 (getaddrinfo.c:605)
==1191==    by 0x4C1BF2B: getaddrinfo (getaddrinfo.c:2306)
==1191==    by 0x3C5A3: socketio_open (socketio_berkeley.c:624)
==1191==    by 0x395DB: xio_open (xio.c:89)
==1191==    by 0x3FA9F: tlsio_openssl_open (tlsio_openssl.c:1243)
==1191==    by 0x395DB: xio_open (xio.c:89)
==1191==    by 0x42373: uws_client_open_async (uws_client.c:1468)
==1191==    by 0x40F7B: wsio_open (wsio.c:509)
==1191==    by 0x395DB: xio_open (xio.c:89)
==1191==    by 0x5CB57: saslclientio_open_async (saslclientio.c:1078)
==1191==    by 0x395DB: xio_open (xio.c:89)
==1191==    by 0x574E7: connection_open (connection.c:1260)
==1191==    by 0x5D25F: session_begin (session.c:760)
==1191== 
==1191== 
==1191== HEAP SUMMARY:
==1191==     in use at exit: 572 bytes in 4 blocks
==1191==   total heap usage: 25,495 allocs, 25,491 frees, 846,817 bytes allocated
==1191== 
==1191== LEAK SUMMARY:
==1191==    definitely lost: 0 bytes in 0 blocks
==1191==    indirectly lost: 0 bytes in 0 blocks
==1191==      possibly lost: 0 bytes in 0 blocks
==1191==    still reachable: 572 bytes in 4 blocks
==1191==         suppressed: 0 bytes in 0 blocks
==1191== Rerun with --leak-check=full to see details of leaked memory
==1191== 
==1191== For counts of detected and suppressed errors, rerun with: -v
==1191== Use --track-origins=yes to see where uninitialised values come from
==1191== ERROR SUMMARY: 3 errors from 1 contexts (suppressed: 21 from 8)
```

**Comment by jspaith** [comment]
Could you also try with `--track-origins=yes -v" additionally?

Looking at the stack for the uninitialized variable, it's hard (not to say impossible) to see how the Azure IoT SDK is responsible for that one.  We're pretty deep down into a loading a library and the last place we interact with the callstack is a simple getaddrinfo().

**Comment by jdswensen** [comment]
Ran with these options: `valgrind --suppressions=./global_valgrind_suppression.supp --num-callers=24 --sigill-diagnostics=no --leak-check=full --show-leak-kinds=all --track-origins=yes -v --leak-resolution=high --log-file="sample-valgrind-imx6ul6.log" ./iothub_client_sample_amqp_websockets`

Log: [sample-valgrind-imx6ul6.log](https://github.com/Azure/azure-iot-sdk-c/files/1687545/sample-valgrind-imx6ul6.log)

I noticed that the libraries valgrind was looking for had CRC mismatches. I updated my build to include all debugging symbols. My image size went from ~200MB to over 1.5GB, but I got better valgrind output.

Better log: [sample-valgrind-imx6ul-all-debug-syms.log](https://github.com/Azure/azure-iot-sdk-c/files/1687678/sample-valgrind-imx6ul-all-debug-syms.log)

**Comment by jspaith** [comment]
Indeed, these are much better logs.  Thanks for staying on this.

To set timing expectation -- I'm about to be out of the office till mid-next week and it's unlikely I'll be able to look into this until then.  My gut still tells me the getaddrinfo() issue is a false-positive or at least out of our SDK's control, but the cert freeing stuff I'll need to research more.

**Comment by jspaith** [comment]
Which version of openssl are you using on your new setup?  I took a look at OpenSSL_1_1_0f as referenced in you Pi setup, but the line #'s don match up here.

**Comment by jdswensen** [comment]
Hmmm...[1.0.2l](http://git.yoctoproject.org/cgit/cgit.cgi/meta-freescale/tree/recipes-connectivity/openssl/openssl-qoriq_1.0.2l.bb) with a bunch of [patches](http://git.yoctoproject.org/cgit/cgit.cgi/meta-freescale/tree/recipes-connectivity/openssl/openssl-qoriq). It looks like there is an updated version available for our toolchain. I can upgrade to [1.1.0g](https://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta/recipes-connectivity/openssl/openssl_1.1.0g.bb?h=rocko) if necessary.

I'm still looking into a way to reproduce this issue on your side. I'm experimenting with virtualizing an ARMv7 architecture via QEMU, but I'm fighting the tools every step of the way. I'll let you know if I get this working. I've got it booting, but I don't have network connectivity yet.

**Comment by jspaith** [comment]
A few things - first, https://stackoverflow.com/questions/3840582/still-reachable-leak-detected-by-valgrind has a good writeup about what the still-reachable means.  It typically reflects one-time allocations where the pointer isn't truly lost.  Our automated tests fwiw only will fail on "definitely lost" which are clearly bugs.

Reading through openssl 1.1.0 code, IoT SDK invoking `OPENSSL_thread_stop` would be the way to cleanup these thread-local variables.  But... the *Notes* in https://www.mankier.com/3/OPENSSL_init_crypto.3ssl list a pretty precise set of reasons why an app would want to call `OPENSSL_thread_stop`.  I think the added risk and complexity of the SDK making this call does not justify avoiding what looks strongly to be a Valgrind/openssl false positive for still-reachable, which isn't on our radar anyway as I said.

Does this make sense?  This has definitely been a research project for me so I'm open to feedback, but pretty sure I'm correct here.

**Comment by jdswensen** [comment]
That makes sense, but I still need a way to cleanup the tool if this is the root cause of the issue I'm having.

I think I also need clarification on the libraries the Azure SDK depends on. 
- Which versions are your automated tests using?
- What are the configuration options? 

I'm asking because I noticed the curl library I'm using was being built with gnutls, so I switched it to build with openssl and reran the same valgrind command. I attached the log output. When I use OpenSSL to build curl, I get a ton of errors and don't even get a valgrind report after completing the run. The SDK might be able to check for platform symbols using [CheckSymbolExists](https://cmake.org/cmake/help/latest/module/CheckSymbolExists.html) and related functions.

I also got the QEMU machine running (and reproducing this issue), so I'll be able to provide you with an environment that can reproduce this issue. I'll work on a write up on getting it setup if that is still needed, but I need to know the library versions the SDK expects.

[sample-valgrind-imx6ul6-curl-openssl.log](https://github.com/Azure/azure-iot-sdk-c/files/1708489/sample-valgrind-imx6ul6-curl-openssl.log)

**Comment by jspaith** [comment]
**First note that the below is subject to change over time, so should be considered a snapshot of our current internal test process.**

We run our tests against Debian 9.2.1.  We don't build the packages directly, but instead apt-get them.

These are what we're running against:
Openssl 1.1.0g
Curl 7.52+
Libc 2.23+

Because you're on Yocto you're not going to get the Debian package manager so we're not going to have a full correspondence.

**Comment by jdswensen** [comment]
@jspaith, thanks for the feedback. I've created a reproducible environment [here](https://github.com/jdswensen/poky). I have a write up for using it, but during my setup I ran across some  issues and related [discussion](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=532059) about valgrind and openssl. I'll start looking into library compile options generated by Yocto. It appears the iMX6 Yocto recipes may do somethings slightly differently than vanilla Yocto. When I moved everything from iMX6 to vanilla, it added some more errors that didn't exist before.

I would still like this issue to remain open while I investigate further.

**Comment by jspaith** [comment]
"I would still like this issue to remain open while I investigate further." 

Agree.  I do like to close out issues when I think they merit closure :), but here I appreciate more investigation is needed on your side and this can be hard.

**Comment by jdswensen** [comment]
It looks like Yocto was building openssl 1.0.1m by default. I forced it to build 1.1.0g instead and all of the crypto mallocs errors went away. I am still getting some errors about `nss_load_library`, see attached.

What is your underlying libc implementation? Are you using glibc, uclibc, etc?

[sample-valgrind-qemuarma9.log](https://github.com/Azure/azure-iot-sdk-c/files/1732086/sample-valgrind-qemuarma9.log)

**Comment by jspaith** [comment]
We're using default libc, which should be glibc unless you've done something to over-ride (and our current cmake don't have this concept).  RE the issue, closest i see is https://stackoverflow.com/questions/41829655/valgrind-memory-leaks-because-of-getaddrinfo stack but it's leak, not uninitialized specific.

Again, this seems to be lower level issue than what IoT SDK itself is doing unless I'm really missing something.

**Comment by jdswensen** [comment]
Agreed. The main issue appears to have been a [bug](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=532059) that was fixed in openssl version 1.1.0. Upgrading removed everything except that final error. I'll see if I can suppress it somehow, but I believe you are correct.

My final request would be that minimum library versions be placed somewhere in the documentation, but preferably enforced through cmake. I can open this as a separate issue if you like.

**Comment by jspaith** [comment]
Yes, please open a separate issue for the min libs and close out this one if you're happy with resolution (at least as far as we can take you).  Thanks

**Comment by jdswensen** [comment]
Sounds good. Thanks for your help!
